csharp/src/Ice/IPEndpoint.cs

```diff
@@ -210,7 +215,61 @@ protected internal override void AppendOptions(StringBuilder sb, char optionSepa
             }
         }
 
-        internal IPEndpoint Clone(ushort port) => port == Port ? this : Clone(Host, port);
+        protected internal override Endpoint Clone(string host) => host == Host ? this : Clone(host, Port);
+
+        protected internal override IEnumerable<Endpoint> ExpandHost()
+        {
+            if (Address == IPAddress.None) // DNS name
+            {
+                try
+                {
+                    return ExpandDnsHostAsync().Result;
+                }
+                catch (AggregateException ex)
+                {
+                    throw ExceptionUtil.Throw(ex.InnerException!);
+                }
+            }
+            else
+            {
+                return ImmutableArray.Create(this);
+            }
+
+            async Task<IEnumerable<Endpoint>> ExpandDnsHostAsync(CancellationToken cancel = default)
+            {
+                try
+                {
+                    IPAddress[] addresses =
+                        await Dns.GetHostAddressesAsync(Host).WaitAsync(cancel).ConfigureAwait(false);
+
+                    return addresses.Select(
+                        address =>
+                        {
+                            IPEndpoint expanded = Clone(address.ToString(), Port);
+                            expanded._address = address;
+                            return expanded;
+                        });
+                }
+                catch (Exception ex)
+                {
+                    throw new DNSException(Host, ex);
+                }
+            }
```

Comment:
Can we just use the sync version `Dns.GetHostAddresses`?

```suggestion
            if (Address == IPAddress.None) // DNS name
            {
                try
                {
                    return Dns.GetHostAddresses(Host).Select(
                        address =>
                        {
                            IPEndpoint expanded = Clone(address.ToString(), Port);
                            expanded._address = address;
                            return expanded;
                        });
                }
                catch (Exception ex)
                {
                    throw new DNSException(Host, ex);
                }
            }
            else
            {
                return ImmutableArray.Create(this);
            }
```

Reply:
Yes, fixed.